dev/Code/CryEngine/CryPhysics/livingentity.cpp

```diff
@@ -340,10 +340,21 @@ float CLivingEntity::UnprojectionNeeded(const Vec3& pos, const quaternionf& qrot
             else if (pentlist[i]->m_parts[0].flags & collider_flags && !IgnoreCollision(m_collisionClass, pentlist[i]->m_collisionClass))
             {
                 CLivingEntity* pent = (CLivingEntity*)pentlist[i];
-                if (fabs_tpl((pos.z + hCollider - hPivot) - (pent->m_pos.z + pent->m_hCyl - pent->m_hPivot)) < newdim.z + pent->m_size.z &&
-                    len2(vector2df(pos) - vector2df(pent->m_pos)) < sqr(newdim.x + pent->m_size.x))
+                float myColliderHalfHeight = newdim.z + (newdim.x * bCapsule);
+                float theirColliderHalfHeight = pent->m_size.z + (pent->m_size.x * pent->m_bUseCapsule);
+                float verticalCenterOfMyCollider = pos.z + hCollider - hPivot;
+                float verticalCenterOfTheirCollider = pent->m_pos.z + pent->m_hCyl - pent->m_hPivot;
+                float distanceBetweenVerticalCentersOfColliders = fabs_tpl(verticalCenterOfMyCollider - verticalCenterOfTheirCollider);
+                Vec2 planarCenterToCenterOfColliders = vector2df(pent->m_pos) - vector2df(pos);
+                float squareDistanceBetweenPlanarCentersOfColliders = len2(planarCenterToCenterOfColliders);
+                if (distanceBetweenVerticalCentersOfColliders < myColliderHalfHeight + theirColliderHalfHeight &&
+                    squareDistanceBetweenPlanarCentersOfColliders < sqr(newdim.x + pent->m_size.x))
                 {
-                    return 1E10;
+                    // Pick the best direction to unproject in the XY plane
+                    Vec2 safeUnprojDir = Vec2(cry_random(-1.0f, 1.0f), cry_random(-1.0f, 1.0f)).GetNormalizedSafe(Vec2(0.0f, 1.0f));
```

Comment:
Can this be simplified to a 1,0 vec2 instead? Is randomization required here?

Reply:
Morning @ranilianaws , 

The rand is required because this code is entered when it was otherwise impossible to work out which direction to unproject, e.g. when 2 objects occupy exactly the same position. Using a fixed value here would mean we are assuming that it is safe to unproject in that direction. If it fails then it would just keep retrying in the same direction until the retry count expired then it would bail out. By using a random direction instead, every retry will try a different direction until it finds one that works. An example use case is when spawning 50 AI characters on the same spot. With the randomisation, they end up spawning in a circle around the spot, but with a fixed direction you would end up with a long line of characters until there was no more room. The random version ensures that all 50 characters are reasonably close to the desired spawn point.

Reply:
Great, thank you!